### PR TITLE
fix(issue-234): newline-only delimiter for RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS in parse_literal_pairs

### LINKED ISSUE #234: [bug]: RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS CSV format breaks when a value contains a comma
### Affected scope

Blueprint source repository

### What happened?

The `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` variable is documented with the format `key=value,key2=value2` (comma-separated key=value pairs). The `parse_literal_pairs()` function in generated consumer scripts splits on every comma via `IFS=',' read -r -a raw_pairs`.

This breaks silently when any value contains a comma. The built-in default for `NUXT_OIDC_TOKEN_KEY` is `data:;base64,bG9jYWwtZGV2LW9pZGMtdG9rLWtleS0zMi1ieXRlcyE=` — a [data URI](https://developer.mozilla.org/en-US/docs/Web/URI/Reference/Schemes/data) that contains a literal comma between the media-type and the base64 payload. When joined with IFS=`,` and then split by the parser, the base64 portion is treated as a separate (malformed) pair, causing `parse_literal_pairs()` to return 1 and the source secret to never be created.

**Failure cascade:**
1. `render_seed_runtime_credentials_literals()` produces a broken CSV containing a split `NUXT_OIDC_TOKEN_KEY` entry.
2. `parse_literal_pairs()` returns exit 1 → `reconcile_eso_runtime_secrets.sh` records a reconcile issue.
3. `runtime-credentials-source` Kubernetes secret is never created.
4. All ExternalSecrets in the `security` and `apps` namespaces remain `NotReady`.
5. `argocd-gitops-repo` secret is never populated → ArgoCD cannot pull manifests → all app deployments fail.

### Reproduction

```bash
# Default value used by provision_deploy_local_marketplace.sh
NUXT_OIDC_TOKEN_KEY="data:;base64,bG9jYWwtZGV2LW9pZGMtdG9rLWtleS0zMi1ieXRlcyE="

# Simulate render_seed_runtime_credentials_literals()
seed_runtime_literal_pairs=(
  "username=local-user"
  "NUXT_OIDC_TOKEN_KEY=${NUXT_OIDC_TOKEN_KEY}"
)
IFS=, csv="${seed_runtime_literal_pairs[*]}"
echo "$csv"
# Output: username=local-user,NUXT_OIDC_TOKEN_KEY=data:;base64,bG9jYWwtZGV2LW9pZGMtdG9rLWtleS0zMi1ieXRlcyE=
# parse_literal_pairs() reads 3 tokens: "username=local-user" | "NUXT_OIDC_TOKEN_KEY=data:;base64" | "bG9j..."
# The third token has no "=" → returns 1
```

### Expected behavior

The format must either:
- Support values that contain commas (via escaping or quoting), **or**
- Switch to a separator that cannot appear in base64 data-URI values (e.g. newline).

A newline-separated format is backward-compatible: `parse_literal_pairs()` already emits newline-separated output. The consumer serializer only needs to change from `IFS=, printf '%s' "${pairs[*]}"` to `printf '%s\n' "${pairs[@]}"`, and the parser needs to accept `\n` as an alternative delimiter.

### Temporary workaround path

`scripts/bin/platform/infra/provision_deploy_local_marketplace.sh` — consumer changed the serializer to newline-separated output and updated `parse_literal_pairs()` in `scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh` to accept both formats.

### Replacement trigger

Blueprint adopts newline-separated (or properly escaped) format for `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` in template docs and generated consumer scripts.

### Workaround review date

2026-07-28

### PR DESCRIPTION
**Pull request by sbonoc**
## Summary

Fixes a silent failure where `parse_literal_pairs()` in `reconcile_eso_runtime_secrets.sh`
split `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` on every comma, causing any value containing
a comma (such as `NUXT_OIDC_TOKEN_KEY=data:;base64,...` data URIs) to be tokenized
into malformed pairs, returning exit code 1, and preventing the `runtime-credentials-source`
Kubernetes secret from ever being created — leaving all ExternalSecrets `NotReady` and
blocking ArgoCD sync for all app namespaces.

Adopts **Option B: newline-only strict** — comma-separated input is rejected outright with
`log_warn` and `record_reconcile_issue` so consumers receive a visible diagnostic regardless
of `RUNTIME_CREDENTIALS_REQUIRED`. No backward compatibility path; consumers must migrate to
`$'key=value\nkey2=value2'` quoting. Migration guide added to docs.

Closes #234.

## Reviewer Package

Full context and evidence: `specs/2026-04-28-issue-234-literal-pairs-newline-format/pr_context.md`

## Key Files

- `scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh` — `parse_literal_pairs()` replacement
- `tests/infra/test_runtime_credentials_eso.py` — 2 new tests (AC-001/AC-002/AC-003), 2 migrated tests
- `docs/platform/consumer/runtime_credentials_eso.md` + bootstrap template — format contract + migration guide
- `docs/platform/consumer/troubleshooting.md` + bootstrap template — seed example updated

## Validation Evidence

- `python3 -m pytest tests/infra/test_runtime_credentials_eso.py -v` — 12/12 PASS
- `make infra-contract-test-fast` — 136/136 PASS
- `make docs-build && make docs-smoke` — PASS
- `make quality-hardening-review` — PASS

## Breaking Change

`RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` now accepts **newline-separated `key=value` pairs only**.
Comma-separated input is rejected with a non-zero exit and `log_warn`.
Consumers who applied the workaround (newline serializer) are unaffected.
Consumers still using comma-separated format must migrate before upgrading.

---
_Generated with [Claude Code](https://claude.ai/claude-code)_

### CODE REVIEW

**Comment by sbonoc** [comment]
SPEC_PRODUCT_READY: approved
ARCHITECTURE_SIGNOFF: approved
SECURITY_SIGNOFF: approved
OPERATIONS_SIGNOFF: approved

**Comment by sbonoc** [comment]
@codex review this PR

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/graph.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    },
+    {
+      "id": "FR-003",
+      "type": "requirement",
+      "summary": "Documentation MUST declare newline-separated as recommended format; comma-separated as legacy"
+    },
+    {
+      "id": "FR-004",
+      "type": "requirement",
+      "summary": "reconcile_eso_runtime_secrets.sh error message MUST reference both accepted formats"
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/graph.json]: Fixed in 69a7b56 — updated FR-003 summary to "Documentation MUST declare newline-separated key=value pairs as the accepted format; comma-separated is no longer accepted" and FR-004 to "reconcile_eso_runtime_secrets.sh error message MUST reference newline-separated format only".

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/context_pack.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Work Item Context Pack
+
+## Context Snapshot
+- Work item: 2026-04-28-issue-234-literal-pairs-newline-format
+- Track: platform
+- SPEC_READY: false
+- ADR path: docs/platform/architecture/decisions/ADR-20260428-issue-234-literal-pairs-newline-format.md
+- ADR status: proposed
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/context_pack.md]: Fixed in 69a7b56 — updated context_pack.md to SPEC_READY: true and ADR status: approved, matching spec.md.

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/evidence_manifest.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "manifest_version": 1,
+  "work_item": "<YYYY-MM-DD>-<work-item-slug>",
+  "generated_by": "spec-evidence-manifest",
+  "generated_at_utc": "",
+  "files": []
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/evidence_manifest.json]: Fixed in 69a7b56 — populated evidence_manifest.json with work_item slug, generated_at_utc, and SHA-256/byte entries for all 14 primary implementation and spec artifact files.

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/spec.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- Local-first exception rationale: none
+
+## Objective
+- Business outcome: Restore correct behavior for `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` when any value contains a comma, eliminating a silent failure cascade that prevents ESO secret reconciliation and breaks all app deployments in affected environments.
+- Success metric: `parse_literal_pairs()` correctly parses a newline-separated entry whose value contains a comma (such as a data URI); comma-separated input is rejected with a visible diagnostic regardless of `RUNTIME_CREDENTIALS_REQUIRED` value.
+
+## Normative Requirements
+
+### Functional Requirements (Normative)
+- FR-001 MUST update `parse_literal_pairs()` in `scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh` to accept ONLY newline-separated `key=value` pairs; EXACTLY ONE splitting strategy SHALL be applied — split on newlines; comma-separated input MUST NOT be supported.
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/spec.md]: Fixed in 69a7b56 — FR-001 wording updated to clarify that EXACTLY ONE parsing strategy applies to accepted input (newlines), and that implementations are permitted to inspect raw input for commas solely to detect and reject deprecated format. Used "are permitted to" rather than "MAY" to satisfy the SDD language policy ("may" is a forbidden ambiguous term in normative sections).

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/plan.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# Implementation Plan
+
+## Implementation Start Gate
+- Implementation tasks MUST remain unchecked until `SPEC_READY=true`.
+- If required inputs are missing, add `BLOCKED_MISSING_INPUTS` in `spec.md` and keep the gate closed.
+
+## Constitution Gates (Pre-Implementation)
+- Simplicity gate:
+  - Change is a direct replacement of the `IFS=',' read -r -a raw_pairs` array split with a `while IFS= read -r pair` newline loop inside `parse_literal_pairs()`. No new abstractions, no detection branch.
+- Anti-abstraction gate:
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/plan.md:10]: Fixed in 69a7b56 — simplicity gate updated to acknowledge the comma-detection heuristic (required for AC-002 rejection) while making clear it does not alter the newline parsing strategy for accepted input.

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/architecture.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+## Bounded Contexts and Responsibilities
+- Parser context (`parse_literal_pairs`): owns delimiter strategy — determine whether input is newline-separated or legacy comma-separated based on presence of `\n`; emit one `key=value` per line; validate `key` and `value` are non-empty.
+- Caller context (`reconcile_eso_runtime_secrets.sh`): reads parsed output line-by-line into `source_literals` array; invokes `apply_optional_module_secret_from_literals`; emits reconcile issue on parse failure.
+
+## High-Level Component Design
+- Domain layer: `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` env var — user-supplied key=value pairs (newline or comma-separated).
+- Application layer: `parse_literal_pairs()` — delimiter detection + per-pair validation → normalized `key=value\n` stream.
+- Infrastructure adapters: `apply_optional_module_secret_from_literals` → `render_optional_module_secret_manifests` → Kubernetes Secret YAML → `kubectl apply`.
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/architecture.md:26]: Fixed in 69a7b56 — removed all stale Option A / "newline or comma-separated" / "delimiter detection" language. Domain layer now reads "newline-separated only"; application layer now reads "format validation + comma-detection heuristic for deprecated-format rejection"; bounded contexts now read "accept only newline-separated input; reject comma-separated input with log_warn diagnostic".

File: specs/2026-04-28-issue-234-literal-pairs-newline-format/architecture.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+- Presentation/API/workflow boundaries: no change — this is an internal shell function.
+
+## Integration and Dependency Edges
+- Upstream dependencies: env var `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` set by consumer scripts or CI/CD.
+- Downstream dependencies: `source_literals` array → `apply_optional_module_secret_from_literals` → `render_optional_module_secret_manifests` → Kubernetes Secret.
+- Data/API/event contracts touched: `RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS` format contract (documented in `runtime_credentials_eso.md`).
+
+## Non-Functional Architecture Notes
+- Security: value content must never be truncated or split — internal commas in values (base64 data URIs, JWTs, connection strings) must pass through verbatim to the base64-encoded Kubernetes secret data field. The `${pair#*=}` greedy strip from the right already handles this correctly for the per-pair value extraction; the fix is entirely in the delimiter detection.
+- Observability: `record_reconcile_issue` call-site is preserved; only the error message string is updated to reference both formats.
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply specs/2026-04-28-issue-234-literal-pairs-newline-format/architecture.md]: Fixed in 69a7b56 — both inaccuracies corrected: (1) `${pair#*=}` now described as "strips the shortest matching prefix from the left through the first =" (not "from the right"); (2) reconcile-issue message now described as "updated to reference the newline-separated-only format" (not "both formats").

File: scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh [hunk 8]
```diff
@@ -243,24 +243,53 @@ should_skip_eso_contract_check() {
...
+        _cp_val="${_cp#*=}"
+        if [[ "$_cp" == *=* && -n "$_cp_val" && "$_cp_key" =~ ^[A-Za-z_][A-Za-z0-9_]*$ ]]; then
+          _detected_csv=true
+          break
+        fi
+      done
+      if [[ "$_detected_csv" == "true" ]]; then
+        log_warn "parse_literal_pairs: input looks like comma-separated format (no longer supported)"
+        log_warn "parse_literal_pairs: use newline-separated key=value pairs (one per line)"
+        return 1
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh:272]: **Disagree — heuristic is intentional; documented workaround covers the edge case.**

The heuristic satisfies the normative AC-002 requirement, which explicitly requires that comma-separated input (`username=dev-user,password=dev-password`) be rejected with a non-zero exit and `log_warn`. Removing the heuristic would make `username=dev-user,password=dev-password` silently accepted as one pair with value `dev-user,password=dev-password`, breaking AC-002.

**On the false-positive scenario (`token=abc,scope=read` as a single pair):** users with a single credential whose value contains `,identifier=value` can avoid the heuristic by appending a trailing newline, which skips the single-line guard:
```bash
export RUNTIME_CREDENTIALS_SOURCE_SECRET_LITERALS=$'token=abc,scope=read\n'
```
The condition `[[ "$literals" != *$'\n'* ]]` is false when a newline is present, so the heuristic block is never entered and the pair is parsed verbatim.

This edge case and workaround will be documented in a follow-up (hardening_review.md Proposal 1 tracks the bash unit-test harness; format edge cases can be addressed there). Credential values in this blueprint's domain (passwords, tokens, URLs, OIDC settings) are unlikely to look like `,identifier=value` in practice.

File: scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh [hunk 9]
```diff
@@ -243,24 +243,53 @@ should_skip_eso_contract_check() {
...
+    local -a _csv_parts
+    IFS=',' read -r -a _csv_parts <<< "$literals"
+    if (( ${#_csv_parts[@]} > 1 )); then
+      local _cp _cp_key _cp_val _detected_csv=false
+      for _cp in "${_csv_parts[@]:1}"; do
+        _cp="$(trim_whitespace "$_cp")"
+        _cp_key="${_cp%%=*}"
+        _cp_val="${_cp#*=}"
+        if [[ "$_cp" == *=* && -n "$_cp_val" && "$_cp_key" =~ ^[A-Za-z_][A-Za-z0-9_]*$ ]]; then
+          _detected_csv=true
```
[the remark this answers is not in the document]
Comment (sbonoc) [reply scripts/bin/platform/auth/reconcile_eso_runtime_secrets.sh:265]: **Disagree — scoping the heuristic to shell-identifier keys is intentional.**

The regex `^[A-Za-z_][A-Za-z0-9_]*$` intentionally covers only shell variable name identifiers (e.g., `username`, `password`, `KEYCLOAK_ADMIN_PASSWORD`). This is the set of keys consumers realistically use in the old comma-separated format, because the old format was typically set via shell variable assignment where key names follow shell-identifier conventions.

Expanding the heuristic to hyphenated keys (e.g., matching `password-token`) would increase the false-positive surface: any value containing `,hyphenated-text=value` (e.g., content-type header strings, MIME type values) would be incorrectly flagged. The identifier-scoped heuristic is the tightest constraint that covers the realistic migration cases without introducing new false positives.

Consumers using hyphenated keys in comma-separated format are outside the expected usage pattern for this env var (which is documented as shell-style key=value pairs for Kubernetes secrets). They will silently get a malformed single pair — which is a visible failure at the kubectl/ESO level, not a silent success.
